mongo/client_bulk_write_models.go

```diff
@@ -0,0 +1,345 @@
+// Copyright (C) MongoDB, Inc. 2024-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package mongo
+
+import (
+	"fmt"
+
+	"go.mongodb.org/mongo-driver/v2/mongo/options"
+)
+
+// ClientWriteModels is a struct that can be used in a client-level BulkWrite operation.
+type ClientWriteModels struct {
+	models []clientWriteModel
+}
+type clientWriteModel struct {
+	namespace string
+	model     interface{}
+}
+
+// AppendInsertOne appends ClientInsertOneModels.
+func (m *ClientWriteModels) AppendInsertOne(database, collection string, models ...*ClientInsertOneModel) *ClientWriteModels {
+	if m == nil {
+		m = &ClientWriteModels{}
+	}
+	for _, model := range models {
+		m.models = append(m.models, clientWriteModel{
+			namespace: fmt.Sprintf("%s.%s", database, collection),
+			model:     model,
+		})
+	}
+	return m
+}
+
+// appendModels is a helper function to append models to ClientWriteModels.
+func appendModels[T ClientUpdateOneModel |
+	ClientUpdateManyModel |
+	ClientReplaceOneModel |
+	ClientDeleteOneModel |
+	ClientDeleteManyModel](
+	m *ClientWriteModels, database, collection string, models []*T) *ClientWriteModels {
+	if m == nil {
+		m = &ClientWriteModels{}
+	}
+	for _, model := range models {
+		m.models = append(m.models, clientWriteModel{
+			namespace: fmt.Sprintf("%s.%s", database, collection),
+			model:     model,
+		})
+	}
+	return m
+}
+
+// AppendUpdateOne appends ClientUpdateOneModels.
+func (m *ClientWriteModels) AppendUpdateOne(database, collection string, models ...*ClientUpdateOneModel) *ClientWriteModels {
```

Comment:
Is there a reason to use per-model-type setters instead of the unexported interface pattern that the existing bulk write models use?

For example:
```go
type ClientWriteModel interface {
	clientWriteModel()
}

type ClientUpdateOneModel struct { ... }

func (ClientUpdateOneModel) clientWriteModel() {}


func (m *ClientWriteModels) Append(database, collection string, models ...ClientWriteModel) *ClientWriteModels {
	m.models = append(...)
	return m
}
```

Reply:
I've revised the `(*Client).BulkWrite()` signature to accept the Namespace-WriteModel pairs in a slice to align with the legacy bulk write API.